flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -369,7 +509,73 @@ public void notifyCheckpointAborted(long checkpointId) throws Exception {
     // Factory function interface
     private interface StateFactory {
         <K, N, SV, S extends State, IS extends S> IS create(
-                InternalKvState<K, N, SV> kvState, KvStateChangeLogger<SV, N> changeLogger)
+                InternalKvState<K, N, SV> kvState,
+                KvStateChangeLogger<SV, N> changeLogger,
+                InternalKeyContext<K> keyContext)
                 throws Exception;
     }
+
+    /**
+     * @param name state name
+     * @param type state type (the only supported type currently are: {@link
+     *     BackendStateType#KEY_VALUE key value}, {@link BackendStateType#PRIORITY_QUEUE priority
+     *     queue})
+     * @return an existing state, i.e. the one that was already created
+     * @throws NoSuchElementException if the state wasn't created
+     * @throws UnsupportedOperationException if state type is not supported
+     */
+    public ChangelogState getExistingState(String name, BackendStateType type)
+            throws NoSuchElementException, UnsupportedOperationException {
+        ChangelogState state;
+        switch (type) {
+            case KEY_VALUE:
+                state = (ChangelogState) keyValueStatesByName.get(name);
```

Comment:
I think there exists something wrong here. `keyValueStatesByName` which is like `AbstractKeyedStateBackend#keyValueStatesByName`, is only updated during `KeyedStateBackend#getOrCreateKeyedState`. 

However, this place is not somewhere we restore and build the keyed state backend but only for the first user access. For RocksDB state backend, `kvStateInformation` is the actual place to store restored state. And for heap keyed state backend, `registeredKVStates` is the place to store restored data.

That is to say, we cannot get existing state when try to apply state changes during restore. Moreover, we also need to consider state migration, the state changes should be appiled by previous state descriptor (mainly the value seralizer) not current state descriptior (as they might not be the same).

Reply:
`ChangelogKeyedStateBackend` differs a bit from other backends here. On recovery,
1. delegated backends are created as usual (with metadata `M1`)
1. `ChangelogBackendLogApplier` reads metadata `M2` in `restoreKvMetaData()` and calls `changelogBackend.getOrCreateKeyedState()` (`M2` replaces `M1` in delegated backend)
1. After that, `ChangelogBackendLogApplier.applyDataChange()` calls `changelogBackend.getExistingState()` to apply data changes
1. User program may call `changelogBackend.getOrCreateKeyedState()` with metadata `M3`

So `changelogBackend.keyValueStatesByName` **is** updated on recovery (in step 2).

---

However, state migration won't currently work as you pointed out.

I see the only way to implement it by using delegating metadata and serializers, similar to delegating functions in this PR:
- on step 2, wrap obtained serializers in `changelogBackend.createInternalState()` and pass this to the `delegatedBackend`
- on step 4, upgrade metadata and replace the wrapped serializers with the upgraded ones

So the metadata of the delegated backend won't be upgraded explicitly, but effictively it will start using the new serializers.

WDYT? (it's probably easier to discuss it offline)

Other approaches don't seem to work:
- If we simply call `delegatedBackend.createInternalState` to upgrade metadata, then already created states won't receive the updated serializers
- If we replace existing internal states with new ones, we can end up with resource leaks (and potentially non-flushed changes)
- (we do need to create internal delegated states during recovery, so that we can apply changes by calling `put`/`add`/`clear` on them)

Reply:
I think we should not call `changelogBackend.getOrCreateKeyedState()` in step2 to put states within delegated backend. And I think delegating state descritptor which includes serializer and function could help here. We would not let delegated backend to call `getOrCreateKeyedState` during restore.

Reply:
> I think we should not call changelogBackend.getOrCreateKeyedState() in step2 to put states within delegated backend.

But how will we apply state changes then? When applying changelog, delegated backend **has** to have internal states created.
Do you mean we should call some other method but still have internal states created in step 2?

> We would not let delegated backend to call getOrCreateKeyedState during restore.

`delegatedBackend.getOrCreateKeyedState` is actually never called. Only `delegatedBackend.createInternalState` is called.

> And I think delegating state descritptor which includes serializer and function could help here.

In the end, internal states have `final` references to serializers. So at least serializers should be delegating.
I think state migration deserves a separate ticket and a PR, which can be applied after this one, WDYT?

Reply:
Let me take `RocksDBKeyedStateBackend` for example.

Operations | Restoring state | User starts to call getOrCreateState for the 1st time
--|------------ | -------------
Normal RocksDBKeyedStateBackend|Restore from checkpoingts, not visiable for users (not stored in `keyValueStatesByName`) | Might trigger state migration and then visiable for users (stored in `keyValueStatesByName`)
Changelog (RocksDB) statebackend|Restore from checkpoingts, visiable for users and then apply state changes ( (stored in `keyValueStatesByName`) | Just get from `keyValueStatesByName` and cannot trigger state migration.

I think the first difference is that shall we make the state visiable for users during restoring for changelog state backend? From my point of view, we could still keep them not visiable so that we can then trigger state migration for the 1st user access.
I agree that we could make state migration as a separate ticket, but I think we should first discuss whether we should make previous state visiable in `keyValueStatesByName` and that could be an easy fix in this PR.

If talking about state migration, I think we could have two choices:
1. Apply state changes during restore. And then migrate state in the 1st user access. This might need data in changelog write twice.
2. Keep the state changes until user access the state during the 1st time. Then we could apply state changes with possible state migration. This could avoid uncessary re-write, but we need to keep those persist state changes.

Maybe we could have basic discussion here for these choices and leave TODO in this PR if necessary.

Reply:
To my understanding, the purpose of `keyValueStatesByName` is to have 
1. exactly-once state table creation (in `createInternalState`)
2. exactly-once metadata update and state migration (in `tryRegisterKvStateInformation`) 

To implement (2), we'll probably introduce a special `changelogBackend.getOrCreateStateForRecovery()`. 
And then both methods will provide (1) - with the help `keyValueStatesByName` map;
and only in the current `getOrCreateState()` will provide (2) - by using some `delegatingSerializers` map for example.
So `keyValueStatesByName` should be updated anyways.

Regarding the approaches, lazy recovery (option 2) seems much more complicated. And because most snapshots I think will have the materialized part (which will be bigger), it's re-write will dominate changelog re-write. So I think it doesn't worth it.

WDYT?

Reply:
I've created FLINK-23143 and will add a corresponding `TODO` in `ChangelogKeyedStateBackend.getOrCreateKeyedState` as agreed offline.

Thanks a lot for bringing this up and for the discussion!